### ISSUE TITLE
Feature Request: Managed Instances

### ISSUE THREAD

**Issue by CamJW101** [MultiMC/MultiMC5#3057]
---------------------------
Hello! I run a semi-large Minecraft community where everyone uses MultiMC to play. We have a server that changes mods, versions, features, etc. fairly often. Right now, the process to update one's instance to connect to our server after a feature change look like this:

1. I make a new instance in MultiMC with the required settings, mods, and version.
2. I export the instance as a ZIP file.
3. I send the ZIP file to a discord server where the players then download it
4. The players make a new instance from that ZIP file and play.
---------------------------
I think this process could be simplified by implementing a feature I'm calling "Managed Instances." Instead of the process above, it would look something like this:

1. I host the zip file somewhere, most likely a web server.
2. The player makes a new instance, and selects a new option to have the settings managed by a remote host.
3. When the player runs the instance on MultiMC, MultiMC downloads the ZIP file from the web server and checks for changes to the settings or files.
4. MultiMC applies the changes and starts the instance.

I think this would be great for communities looking to streamline their custom feature distributions. It would also make it much easier to identify bugs and distribute bug fixes that would affect everyone. Of course the "Managed Instance" wouldn't affect player specific settings like RAM and Credentials.

Let me know what you think! I think this would be pretty feasible to add to the existing program.

**Comment by CamJW101** [comment]
I would appreciate constructive criticism and reasons as to why this is not a feature worth adding instead of 👎s, but who am I to judge?

**Comment by v3Monsta** [comment]
If I had to guess as to why people are wary about this idea, it's because there is a non-zero chance that the remote host could potentially install a malicious instance. I have no clue how an instance could be malicious-- maybe it's as simple as distributing something malicious alongside the actual instance, but there is no denying that it's possible.

What do I know though, that's just my initial reaction and I have no idea how this attack would work. I suppose you could also say that MMC already downloads stuff from remote hosts (see FTB Legacy), but those are one-off installs and isn't some random host somewhere doing whatever it wants.

**Comment by CamJW101** [comment]
That is true. I don't know how you would go about checking for malicious content. I guess there could be disclaimers as well warning people to make sure they trust the external host. That OR the distribution could be controlled by MultiMC, and settings/instances could be shared with codes generated in the Application.

**Comment by comp500** [comment]
Modpack updating would certainly be beneficial, but I think it would be preferable for there to be an Update button rather than the update occurring whenever the pack is started. This conveys the intent of the action of the button better, and means that malicious updates are easier to spot.

**Comment by Xalaxis** [comment]
It's worth noting that in the scenario @CamJW101 proposes (creating a zip instance every time there is an update and distributing it) there would be the same opportunity for a malicious file to be included in the zip whether it is distributed manually or automatically.  It's also worth noting that MultiMC already has functionality for downloading a ZIP instance one-time at creation:

![image](https://user-images.githubusercontent.com/12937964/77827013-2d7d8080-710b-11ea-9136-046ce3385c8c.png)

So I think adding a warning message with the option for the user to tick "I trust this source" should be enough to allay any concerns, it's the responsibility of users (and in the case of CraftForge, Curse) to accept the risk for malware concerns.  This feature would be extremely useful if implemented and is distinct from #2640 because it's more to do with "user managed" packs rather than those hosted on a third-party service.

**Comment by Meegooo** [comment]
Now that Twitch pack search is there, this would be a nice feature to update those modpacks.

**Comment by Xalaxis** [comment]
#2640 is for the update of "third-party" modpacks.  This is more for personal ZIP packs.

**Comment by bolte-io** [comment]
What's needed is an API like Technic Solder, but uses Curseforge as it's mod repository. IMO.
Then MultiMC could just support it like it already does with the current Solder.
Packs could then be created online on the suggested platform and would update on each run by removing/adding mods.

**Comment by kb-1000** [comment]
MultiMC does not update Technic Solder packs on each run as of right now either.
Besides that, the Technic Platform is generally a mess, IMO.

**Comment by Xastabus** [comment]
Even if this was a function end-users had to manually initiate it would still be a great benefit.

I started running a small community server with a custom modpack.  My users are smart people, but they are not A+ certified techs.  The task of manually maintaining a modpack with around 180+ mods can be daunting for them even while I'm walking them through it.

Some method, any method that would allow server ops and modpack authors a way to provide MultiMC modpack update file and have MultiMC do the technical work of any version changes and mod changes included in the zip is desperately needed.  Even if the update zip just contains a change manifest and the mods neccessary to do the work.

MultiMC already allows importing custom modpacks via zip file. A feature to update already imported custom modpacks really shouldn't raise new concerns not already considered when the decision was made to allow importing custom modpacks in the first place.

**Comment by AxoCode** [comment]
What about the remote URL is at the instance.cfg? And then can be set at Settings > Miscellaneous
Maybe also support git repo?
So then we can choose the branch (just like multimc update channel)

**Comment by bolte-io** [comment]
You an achieve something like this with Packwiz, with a little node scripting I can automate adding/removal of mods with it.
You then distribute the instance zip with the Packwiz bootstrapper. It will then auto-update on each run. The pack manifest files can be branched to Github pages to be internet accessible.

Still, would be nice for Multi-MC to have it's own implementation. Or maybe bake Pazkwiz in?

**Comment by Cisien** [comment]
I'd love to see the ability to add a custom feed to the list on the left of the new instance screen, complete with options around update options (interval, whether to check or be automatic, etc), signing certificate options, etc.

**Comment by AxoCode** [comment]
> You an achieve something like this with Packwiz, with a little node scripting I can automate adding/removal of mods with it.
> You then distribute the instance zip with the Packwiz bootstrapper. It will then auto-update on each run. The pack manifest files > can be branched to Github pages to be internet accessible.
> 
> Still, would be nice for Multi-MC to have it's own implementation. Or maybe bake Pazkwiz in?

oh, i dont know packwiz exist. Thanks

**Comment by kitsushadow** [comment]
This feature would be ideal for private packs. 
There is already a create instance from URL option. If an update instance feature was included the base implementation could "reimport" the pack, leave files that are already there, minus mods, and replacing files that match. It's the removing if files which is not so easy but that's not all to difficult and could be the second iteration of implementation.

**Comment by SeymourGx** [comment]
Bit of a necropost at this point, but it really would be useful to have an authoritative instance template in the cloud that users could mirror and download from directly. Only feature I feel is particularly missing from MultiMC. Especially when playing with those less tech-savvy, being able to manage their pack remotely would be a godsend rather than walking them all through how to download mod updates, delete the old files, and tweak config files. Giving them a brand new instance every time works, but that still runs the issue of teaching these players every time how to move over their own configs or what have you. This feature just ticks so many boxes!

**Comment by mk-pmb** [comment]
I'd like an option to configure a custom "external updater command" for my instance that runs before MultiMC's integrated update checks. The external updater will be responsible for obtaining the new files and validating software signatures. The burden of figuring out the details can thus be delegated to 3rd party projects.

The external updater command would be run with current working directory set to the instance path.

I have some ideas how MultiMC could read progress information from the external updater, but for a start we can make it the external updater's responsibility to create its own window to display progress information.

**Comment by Nox13last** [comment]
I love looking at this ticket, seeing 32 👍, and then going back and reading that knob comment about how "this is on nobody's radar.".

Anyway, here's my idea for an implementation:
1. When adding an instance, instead of linking to a ZIP, link to a metafile. The sort of user who would be using it would have a hosting option already, be it Dropbox or Github or whatever. Inside that metafile, the owner writes:
- The current version of the modpack.
- The URL of the repository to grab the modpack ZIP from.
- Any other metadata, the owner's name/Discord/contacts/whatever, JVM arguments, resources, and so on.
2. When launching the instance, have MultiMC very quickly retrieve this metafile and check the version-field. If it this a subsequent launch, and this number is higher than that of the last retrieval, redownload the ZIP from the URL-field.
3. If the number is the same, or the metafile can't be acquired, then just launch what is already there.

**Comment by mk-pmb** [comment]
Potential extensions:
* Multiple URLs for the ZIP files so we can have mirror/fallback servers.
* The manifest could declare version steps: old version, new version, differential ZIP file URL(s).
  * This would allow for quicker download of incremental updates.
  * Not sure if we should want a way to also delete files in an update.
  * Giving both old and new version (instead of having implicit expectations) allows for very flexible shortcuts in the update graph: If an intermediate version introduced a lot of helper mods that later became redundant, the publisher could create such a shortcut ZIP and the updater could entirely skip those helper mods.
